File: velocity-sdk-go/autoapply/autoapply.go
// Package autoapply provides auto-registration for workflows and activities in the VELOCITY Go SDK.
//
// This package enables annotation-driven workflow and activity registration using Go's init()
// pattern. When a workflow or activity is registered via RegisterWorkflow or RegisterActivity,
// it is stored in a global registry. The Worker discovers all registered handlers at startup.
//
// Example:
//
//	func init() {
//	    autoapply.RegisterWorkflow("OrderWorkflow", &OrderWorkflow{})
//	    autoapply.RegisterActivity("process_payment", ProcessPayment)
//	}
//
//	type OrderWorkflow struct{}
//	func (w *OrderWorkflow) Run(ctx WorkflowContext, orderID string) (string, error) {
//	    return ctx.ExecuteActivity("process_payment", orderID)
//	}
//
//	func ProcessPayment(orderID string) (string, error) {
//	    return "charged", nil
//	}
package autoapply

import (
	"fmt"
	"sync"
)

// WorkflowHandler is the interface that all workflow implementations must satisfy.
type WorkflowHandler interface {
	// Run executes the workflow with the given context and input.
	Run(ctx interface{}, input interface{}) (interface{}, error)
}

// ActivityHandler is a function that executes an activity.
type ActivityHandler func(ctx interface{}, input interface{}) (interface{}, error)

var (
	workflowRegistry = make(map[string]func() WorkflowHandler)
	activityRegistry = make(map[string]ActivityHandler)
	registryMutex    sync.RWMutex
)

// RegisterWorkflow registers a workflow factory in the global registry.
//
// The factory function should return a new instance of the workflow handler.
// This allows the Worker to create fresh instances for each workflow execution.
//
// Example:
//
//	autoapply.RegisterWorkflow("OrderWorkflow", func() autoapply.WorkflowHandler {
//	    return &OrderWorkflow{}
//	})
func RegisterWorkflow(workflowType string, factory func() WorkflowHandler) {
	registryMutex.Lock()
	defer registryMutex.Unlock()
	workflowRegistry[workflowType] = factory
}

// RegisterActivity registers an activity handler in the global registry.
//
// Example:
//
//	autoapply.RegisterActivity("process_payment", func(ctx interface{}, input interface{}) (interface{}, error) {
//	    return ProcessPayment(input.(string))
//	})
func RegisterActivity(activityName string, handler ActivityHandler) {
	registryMutex.Lock()
	defer registryMutex.Unlock()
	activityRegistry[activityName] = handler
}

// GetRegisteredWorkflows returns a copy of all registered workflow types.
func GetRegisteredWorkflows() map[string]func() WorkflowHandler {
	registryMutex.RLock()
	defer registryMutex.RUnlock()
	result := make(map[string]func() WorkflowHandler, len(workflowRegistry))
	for k, v := range workflowRegistry {
		result[k] = v
	}
	return result
}

// GetRegisteredActivities returns a copy of all registered activity names.
func GetRegisteredActivities() map[string]ActivityHandler {
	registryMutex.RLock()
	defer registryMutex.RUnlock()
	result := make(map[string]ActivityHandler, len(activityRegistry))
	for k, v := range activityRegistry {
		result[k] = v
	}
	return result
}

// CreateWorkflow creates a new instance of a registered workflow by type.
func CreateWorkflow(workflowType string) (WorkflowHandler, error) {
	registryMutex.RLock()
	defer registryMutex.RUnlock()
	factory, ok := workflowRegistry[workflowType]
	if !ok {
		return nil, fmt.Errorf("no workflow registered for type: %s", workflowType)
	}
	return factory(), nil
}

// GetActivity returns a registered activity handler by name.
func GetActivity(activityName string) (ActivityHandler, error) {
	registryMutex.RLock()
	defer registryMutex.RUnlock()
	handler, ok := activityRegistry[activityName]
	if !ok {
		return nil, fmt.Errorf("no activity registered for name: %s", activityName)
	}
	return handler, nil
}

// ClearRegistries clears both workflow and activity registries (useful for testing).
func ClearRegistries() {
	registryMutex.Lock()
	defer registryMutex.Unlock()
	workflowRegistry = make(map[string]func() WorkflowHandler)
	activityRegistry = make(map[string]ActivityHandler)
}

// WorkflowCount returns the number of registered workflows.
func WorkflowCount() int {
	registryMutex.RLock()
	defer registryMutex.RUnlock()
	return len(workflowRegistry)
}

// ActivityCount returns the number of registered activities.
func ActivityCount() int {
	registryMutex.RLock()
	defer registryMutex.RUnlock()
	return len(activityRegistry)
}

File: sdk/rust/src/auto_apply.rs
//! Auto-apply workflow and activity registration for the VELOCITY Rust SDK.
//!
//! Provides a trait-based registration system where workflow and activity
//! implementations register themselves into a global registry. The Worker
//! discovers all registered handlers at startup — no manual wiring needed.
//!
//! # Example
//!
//! ```rust,no_run
//! use velocity_sdk::auto_apply::{register_workflow, register_activity, WorkflowHandler, ActivityHandler};
//!
//! struct OrderWorkflow;
//! impl WorkflowHandler for OrderWorkflow {
//!     fn workflow_type() -> &'static str { "OrderWorkflow" }
//!     // ... execute implementation
//! }
//!
//! fn process_payment(activity_id: &str, input: &[u8]) -> Vec<u8> {
//!     b"charged".to_vec()
//! }
//!
//! // Register at startup (or use the inventory crate for link-time registration)
//! register_workflow::<OrderWorkflow>();
//! register_activity("process_payment", process_payment);
//! ```

use std::collections::HashMap;
use std::sync::{Arc, Mutex, OnceLock};

// ─── Handler Traits ──────────────────────────────────────────────────────────

/// Trait for workflow implementations.
///
/// Implement this trait on a struct to make it discoverable by the Worker.
/// The `workflow_type()` method returns the type name used for dispatch.
pub trait WorkflowHandler: Send + Sync + 'static {
    /// The workflow type name used for dispatch (e.g., "OrderWorkflow").
    fn workflow_type() -> &'static str;

    /// Execute the workflow with the given context and input.
    fn execute(
        &self,
        ctx: &WorkflowHandlerContext,
        input: &[u8],
    ) -> Result<Vec<u8>, Box<dyn std::error::Error + Send + Sync>>;
}

/// Trait for activity implementations.
///
/// Activities are plain functions that perform non-deterministic work.
pub type ActivityHandler = Arc<dyn Fn(&ActivityHandlerContext, &[u8]) -> Result<Vec<u8>, Box<dyn std::error::Error + Send + Sync>> + Send + Sync>;

// ─── Context Types ───────────────────────────────────────────────────────────

/// Context available inside workflow execution.
#[derive(Debug, Clone)]
pub struct WorkflowHandlerContext {
    pub workflow_key: u64,
    pub workflow_id: String,
    pub run_id: String,
    pub workflow_type: String,
    pub task_queue: String,
    pub current_step: u32,
}

/// Context available inside activity execution.
#[derive(Debug, Clone)]
pub struct ActivityHandlerContext {
    pub activity_type: String,
    pub activity_id: String,
    pub attempt: u32,
}

// ─── Global Registry ─────────────────────────────────────────────────────────

/// Factory function that creates a workflow handler instance.
type WorkflowFactory = Arc<dyn Fn() -> Box<dyn WorkflowHandlerInstance> + Send + Sync>;

/// Object-safe version of WorkflowHandler for dynamic dispatch.
pub trait WorkflowHandlerInstance: Send + Sync {
    fn execute(
        &self,
        ctx: &WorkflowHandlerContext,
        input: &[u8],
    ) -> Result<Vec<u8>, Box<dyn std::error::Error + Send + Sync>>;
}

/// Blanket implementation: any WorkflowHandler is also a WorkflowHandlerInstance.
impl<T: WorkflowHandler> WorkflowHandlerInstance for T {
    fn execute(
        &self,
        ctx: &WorkflowHandlerContext,
        input: &[u8],
    ) -> Result<Vec<u8>, Box<dyn std::error::Error + Send + Sync>> {
        WorkflowHandler::execute(self, ctx, input)
    }
}

struct Registry {
    workflows: HashMap<String, WorkflowFactory>,
    activities: HashMap<String, ActivityHandler>,
}

static REGISTRY: OnceLock<Mutex<Registry>> = OnceLock::new();

fn registry() -> &'static Mutex<Registry> {
    REGISTRY.get_or_init(|| {
        Mutex::new(Registry {
            workflows: HashMap::new(),
            activities: HashMap::new(),
        })
    })
}

// ─── Registration Functions ──────────────────────────────────────────────────

/// Register a workflow type in the global registry.
///
/// Call this at startup (e.g., in `main()`) or use the `inventory` crate
/// for link-time auto-registration.
pub fn register_workflow<T: WorkflowHandler>() {
    let factory: WorkflowFactory = Arc::new(|| Box::new(T) as Box<dyn WorkflowHandlerInstance>);
    let mut reg = registry().lock().unwrap();
    reg.workflows.insert(T::workflow_type().to_string(), factory);
}

/// Register an activity function in the global registry.
pub fn register_activity(name: &str, handler: impl Fn(&ActivityHandlerContext, &[u8]) -> Result<Vec<u8>, Box<dyn std::error::Error + Send + Sync>> + Send + Sync + 'static) {
    let mut reg = registry().lock().unwrap();
    reg.activities.insert(name.to_string(), Arc::new(handler));
}

/// Get all registered workflow type names.
pub fn registered_workflow_types() -> Vec<String> {
    let reg = registry().lock().unwrap();
    reg.workflows.keys().cloned().collect()
}

/// Get all registered activity names.
pub fn registered_activity_names() -> Vec<String> {
    let reg = registry().lock().unwrap();
    reg.activities.keys().cloned().collect()
}

/// Create a workflow handler instance by type name.
pub fn create_workflow(workflow_type: &str) -> Option<Box<dyn WorkflowHandlerInstance>> {
    let reg = registry().lock().unwrap();
    reg.workflows.get(workflow_type).map(|factory| factory())
}

/// Get an activity handler by name.
pub fn get_activity(name: &str) -> Option<ActivityHandler> {
    let reg = registry().lock().unwrap();
    reg.activities.get(name).cloned()
}

/// Clear both registries (useful for testing).
pub fn clear_registries() {
    let mut reg = registry().lock().unwrap();
    reg.workflows.clear();
    reg.activities.clear();
}

/// Count of registered workflows.
pub fn workflow_count() -> usize {
    let reg = registry().lock().unwrap();
    reg.workflows.len()
}

/// Count of registered activities.
pub fn activity_count() -> usize {
    let reg = registry().lock().unwrap();
    reg.activities.len()
}
